rust/ast/src/associated_data.rs

```diff
@@ -0,0 +1,126 @@
+use crate::source_location_accessor::SourceLocationAccessor;
+use crate::type_id::{NodeTypeId, NodeTypeIdAccessor};
+use crate::SourceLocation;
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+struct Key {
+    type_id: NodeTypeId,
+    loc: SourceLocation,
+}
+
+impl Key {
+    fn new<NodeT>(node: &NodeT) -> Self
+    where
+        NodeT: SourceLocationAccessor + NodeTypeIdAccessor,
+    {
+        Self {
+            type_id: node.get_type_id(),
+            loc: node.get_loc(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Item<ValueT> {
+    key: Key,
+
+    // Using `Option` to take this value out, and also populate the value later.
+    value: Option<ValueT>,
+}
+
+impl<ValueT> Item<ValueT> {
+    fn new(key: Key, value: Option<ValueT>) -> Self {
+        Self { key, value }
+    }
+}
+
+type ItemIndex = usize;
+
+#[derive(Debug)]
+pub struct AssociatedDataItemIndex {
+    index: ItemIndex,
+}
+impl AssociatedDataItemIndex {
+    fn new(index: ItemIndex) -> Self {
+        Self { index }
+    }
+}
+
+// A map from AST node to `T`, to assocaite extra data to AST node.
+#[derive(Debug)]
+pub struct AssociatedData<ValueT> {
+    next_index: ItemIndex,
+    items: Vec<Item<ValueT>>,
+    map: HashMap<Key, ItemIndex>,
```

Comment:
By default, Rust's `HashMap` uses a cryptographically secure hash function. It is a bit slow. It can be changed.

Since we are writing a parser that operates on untrusted input, we should probably keep the default, to prevent malicious input from DoS-ing us with hash collisions.